diff -urN UXP-RELBASE_20210817/layout/base/nsCSSFrameConstructor.cpp xulrunner-source/layout/base/nsCSSFrameConstructor.cpp
--- UXP-RELBASE_20210817/layout/base/nsCSSFrameConstructor.cpp	2021-08-15 16:31:05.000000000 +0200
+++ xulrunner-source/layout/base/nsCSSFrameConstructor.cpp	2021-08-18 12:05:41.576894555 +0200
@@ -921,7 +921,7 @@
     if (mCurrentPendingBindingInsertionPoint) {
       mCurrentPendingBindingInsertionPoint->setPrevious(aPendingBinding);
     } else {
-      mPendingBindings.insertBack(aPendingBinding);
+      mPendingBindings.insertFront(aPendingBinding);
     }
   }
 
@@ -6441,7 +6441,7 @@
     // If either of these are null, no sane comparison can be made.
     return false;
   }
-  
+
   nsIFrame* parentFrame = aSibling->GetParent();
   nsIAtom* parentType = parentFrame->GetType();
 
